### PR TITLE
proj: support for conan v2

### PR DESCRIPTION
**Pull request by paulharris**
A nice cmake project, but this is consumed a lot of other difference recipes.
I wonder about bin paths, virtual run envs and other such things will work properly with this new recipe.
How best to test?

closes #14090 
closes https://github.com/conan-io/conan-center-index/issues/13422

### CODE REVIEW

**Comment by prince-chrismc** [comment]
Seems like the deps need some love

**Comment by paulharris** [comment]
Question: VTK imports proj as 
`find_package(LibPROJ (perhaps flags here))`
and `LibPROJ::LibPROJ`
and checks `LibPROJ_MAJOR_VERSION`

The recipe exports proj::proj and proj_MAJOR_VERSION

What is the best way to create an alias for this?  Or should we patch VTK instead?

**Comment by paulharris** [comment]
Can I get this recipe looked at please @prince-chrismc ?
I think the only thing missing is stdcpp_library which requires 1.54 ...

File: recipes/proj/all/test_v1_package/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+    def build_requirements(self):
+        if self.settings.os == "Macos" and self.settings.arch == "armv8":
+            # Workaround for CMake bug with error message:
+            # Attempting to use @rpath without CMAKE_SHARED_LIBRARY_RUNTIME_C_FLAG being
+            # set. This could be because you are using a Mac OS X version less than 10.5
+            # or because CMake's platform configuration is corrupt.
+            # FIXME: Remove once CMake on macOS/M1 CI runners is upgraded.
+            self.build_requires("cmake/3.22.0")
+
```
Comment (SpaceIm) [inline recipes/proj/all/test_v1_package/conanfile.py:17]:
```suggestion
```

File: recipes/proj/all/test_v1_package/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+conan_basic_setup(TARGETS)
+
+add_executable(test_package ../test_package/test_package.c)
+if(PROJ_VERSION_GE_7)
+  find_package(proj REQUIRED CONFIG)
+  target_link_libraries(test_package PROJ::proj)
+else()
+  find_package(proj4 REQUIRED CONFIG)
+  target_link_libraries(test_package PROJ4::proj)
+endif()
```
Comment (SpaceIm) [inline recipes/proj/all/test_v1_package/CMakeLists.txt:14]: cmake_minimum_required(VERSION 3.1)
project(test_package)

include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
conan_basic_setup(TARGETS)

add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/../test_package
                 ${CMAKE_CURRENT_BINARY_DIR}/test_package)

Comment (SpaceIm) [inline recipes/proj/all/test_v1_package/CMakeLists.txt:14]:
```suggestion
cmake_minimum_required(VERSION 3.1)
project(test_package)

include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
conan_basic_setup(TARGETS)

add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/../test_package
                 ${CMAKE_CURRENT_BINARY_DIR}/test_package)
```

File: recipes/proj/all/test_package/conanfile.py [hunk 3]
```diff
@@ -1,27 +1,33 @@
...
-            self.build_requires("cmake/3.22.0")
+    def requirements(self):
+        self.requires(self.tested_reference_str)
+
+    def layout(self):
+        cmake_layout(self)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["PROJ_VERSION_GE_7"] = Version(self.dependencies[self.tested_reference_str].ref.version) >= "7.0.0"
```
Comment (SpaceIm) [inline recipes/proj/all/test_package/conanfile.py:22]:
```suggestion
        tc.variables["PROJ_VERSION_GE_7"] = Version(self.dependencies["proj"].ref.version) >= "7.0.0"
```

File: recipes/proj/all/conanfile.py [hunk 4]
```diff
@@ -1,14 +1,22 @@
-from conans import ConanFile, CMake, tools
+from conan import ConanFile
+from conan.tools.apple import is_apple_os
+from conan.tools.build import stdcpp_library
```
Comment (SpaceIm) [inline recipes/proj/all/conanfile.py:3]:
```suggestion
from conans.tools import stdcpp_library
```

File: recipes/proj/all/conanfile.py [hunk 5]
```diff
@@ -30,162 +38,157 @@ class ProjConan(ConanFile):
...
                 else:
                     cmake_sqlite_call = "generate_proj_db.cmake"
                     pattern = "\"${EXE_SQLITE3}\""
-                lib_paths = self.deps_cpp_info["sqlite3"].lib_paths
-                tools.replace_in_file(
-                    os.path.join(self._source_subfolder, "data", cmake_sqlite_call),
-                    "COMMAND {}".format(pattern),
-                    "COMMAND ${{CMAKE_COMMAND}} -E env \"DYLD_LIBRARY_PATH={}\" {}".format(
-                        ":".join(lib_paths), pattern
+                lib_paths = self.dependencies["sqlite3"].cpp_info.libdirs
```
Comment (SpaceIm) [inline recipes/proj/all/conanfile.py:113]:
```suggestion
                lib_paths = self.dependencies.build["sqlite3"].cpp_info.libdirs
```

and therefore you can remove the condition `if not hasattr(self, "settings_build"):` above

File: recipes/proj/all/conanfile.py (hunk 4 above)
Comment (prince-chrismc) [reply recipes/proj/all/conanfile.py:3]: 👍

File: recipes/proj/all/conanfile.py [hunk 6]
```diff
@@ -30,162 +38,137 @@ class ProjConan(ConanFile):
...
-                    cmake_sqlite_call = "generate_proj_db.cmake"
-                    pattern = "\"${EXE_SQLITE3}\""
-                lib_paths = self.deps_cpp_info["sqlite3"].lib_paths
-                tools.replace_in_file(
-                    os.path.join(self._source_subfolder, "data", cmake_sqlite_call),
-                    "COMMAND {}".format(pattern),
-                    "COMMAND ${{CMAKE_COMMAND}} -E env \"DYLD_LIBRARY_PATH={}\" {}".format(
-                        ":".join(lib_paths), pattern
-                    ),
-                )
```
Comment (SpaceIm) [inline recipes/proj/all/conanfile.py:128]: No it's not what I meant. I said that the block under `if not hasattr(self, "settings_build")` can be applied unconditionally:

```python
        # Agressive workaround against SIP on macOS, to handle sqlite3 executable
        # linked to shared sqlite3 lib
        if is_apple_os(self):
            if Version(self.version) < "8.1.0":
                cmake_sqlite_call = "CMakeLists.txt"
                pattern = "${EXE_SQLITE3}"
            else:
                cmake_sqlite_call = "generate_proj_db.cmake"
                pattern = "\"${EXE_SQLITE3}\""
            lib_paths = self.dependencies["sqlite3"].cpp_info.libdirs
            replace_in_file(
                self,
                os.path.join(self.source_folder, "data", cmake_sqlite_call),
                "COMMAND {}".format(pattern),
                "COMMAND ${{CMAKE_COMMAND}} -E env \"DYLD_LIBRARY_PATH={}\" {}".format(
                    ":".join(lib_paths), pattern
                ),
            )
```

Comment (paulharris) [reply recipes/proj/all/conanfile.py:128]: ah whoops, thanks

**Review by SpaceIm — CHANGES_REQUESTED** [review]
sqlite3 must be added to build requirements unconditionally now.

File: recipes/proj/all/conanfile.py [hunk 7]
```diff
@@ -30,162 +38,153 @@ class ProjConan(ConanFile):
...
+            self.options.rm_safe("fPIC")
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
 
     def requirements(self):
-        self.requires("nlohmann_json/3.10.5")
-        self.requires("sqlite3/3.38.5")
+        self.requires("nlohmann_json/3.11.2")
+        self.requires("sqlite3/3.39.4")
```
Comment (SpaceIm) [inline recipes/proj/all/conanfile.py:64]:
```suggestion
        self.requires("sqlite3/3.40.0")
```

File: recipes/proj/all/conanfile.py [hunk 8]
```diff
@@ -30,162 +38,153 @@ class ProjConan(ConanFile):
...
-        if tools.Version(self.version) < "7.0.0":
-            self._cmake.definitions["PROJ_TESTS"] = False
-            self._cmake.definitions["BUILD_LIBPROJ_SHARED"] = self.options.shared
-            self._cmake.definitions["ENABLE_LTO"] = False
-            self._cmake.definitions["JNI_SUPPORT"] = False
+        if Version(self.version) < "8.1.0":
+            rmdir(self, os.path.join(self.source_folder, "include", "proj", "internal", "nlohmann"))
+
+
+    def generate(self):
```
Comment (SpaceIm) [inline recipes/proj/all/conanfile.py:122]: move generate() before build() please

File: recipes/proj/all/conanfile.py [hunk 9]
```diff
@@ -30,162 +38,153 @@ class ProjConan(ConanFile):
...
-    def _source_subfolder(self):
-        return "source_subfolder"
-
-    @property
-    def _is_msvc(self):
-        return str(self.settings.compiler) in ["Visual Studio", "msvc"]
-
     @property
     def _settings_build(self):
         return getattr(self, "settings_build", self.settings)
```
Comment (SpaceIm) [inline recipes/proj/all/conanfile.py:44]:
```suggestion
```

can be removed now

File: recipes/proj/all/test_package/conanfile.py [hunk 10]
```diff
@@ -1,27 +1,33 @@
-from conans import ConanFile, CMake, tools
+from conan import ConanFile
+from conan.tools.build import can_run
+from conan.tools.cmake import cmake_layout, CMake, CMakeToolchain
+from conan.tools.scm import Version
 import os
 
 
+# It will become the standard on Conan 2.x
```
Comment (SpaceIm) [inline recipes/proj/all/test_package/conanfile.py:8]:
```suggestion
```

File: recipes/proj/all/conanfile.py (hunk 8 above)
Comment (paulharris) [reply recipes/proj/all/conanfile.py:122]: Ok, I left it in the original order to make the diff easier to read.  Time to move!

Comment (paulharris) [reply recipes/proj/all/conanfile.py:117]: I feel like _patch_sources() should be folded into build().   Do you like to do that?

File: recipes/proj/all/conanfile.py [hunk 11]
```diff
@@ -30,162 +38,148 @@ class ProjConan(ConanFile):
...
-                    ),
-                )
+        if is_apple_os(self):
+            if Version(self.version) < "8.1.0":
+                cmake_sqlite_call = "CMakeLists.txt"
+                pattern = "${EXE_SQLITE3}"
+            else:
+                cmake_sqlite_call = "generate_proj_db.cmake"
+                pattern = "\"${EXE_SQLITE3}\""
+            lib_paths = self.dependencies.build["sqlite3"].cpp_info.libdirs
```
Comment (SpaceIm) [inline recipes/proj/all/conanfile.py:137]:
```suggestion
            if hasattr(self, "settings_build"):
                lib_paths = self.dependencies.build["sqlite3"].cpp_info.libdirs
            else:
                lib_paths = self.dependencies["sqlite3"].cpp_info.libdirs
```

**Comment by SpaceIm** [comment]
@paulharris please consider https://github.com/paulharris/conan-center-index/pull/4

**Comment by paulharris** [comment]
I don't cross-compile so I'll trust that you know what you are doing!

File: recipes/proj/all/patches/0001-use-cmake-targets-9.0.1.patch [hunk 12]
```diff
@@ -1,27 +1,30 @@
 --- a/src/lib_proj.cmake
 +++ b/src/lib_proj.cmake
-@@ -444,8 +444,7 @@ if(USE_THREAD AND Threads_FOUND AND CMAKE_USE_PTHREADS_INIT)
+@@ -444,8 +444,8 @@ if(USE_THREAD AND Threads_FOUND AND CMAKE_USE_PTHREADS_INIT)
```
Comment (SpaceIm) [inline recipes/proj/all/patches/0001-use-cmake-targets-9.0.1.patch:3]:
```suggestion
@@ -444,8 +444,7 @@ if(USE_THREAD AND Threads_FOUND AND CMAKE_USE_PTHREADS_INIT)
```

File: recipes/proj/all/patches/0001-use-cmake-targets-9.0.1.patch [hunk 13]
```diff
@@ -1,27 +1,30 @@
...
    target_link_libraries(proj PRIVATE ${CMAKE_THREAD_LIBS_INIT})
  endif()
  
 -target_include_directories(proj PRIVATE ${SQLITE3_INCLUDE_DIR})
 -target_link_libraries(proj PRIVATE ${SQLITE3_LIBRARY})
-+target_link_libraries(proj PRIVATE CONAN_PKG::sqlite3)
++find_package(SQLite3 REQUIRED)
```
Comment (SpaceIm) [inline recipes/proj/all/patches/0001-use-cmake-targets-9.0.1.patch:9]:
```suggestion
```

File: recipes/proj/all/patches/0001-use-cmake-targets-9.0.1.patch [hunk 14]
```diff
@@ -1,27 +1,30 @@
...
++target_link_libraries(proj PRIVATE SQLite::SQLite3)
  
  if(NLOHMANN_JSON STREQUAL "external")
    target_compile_definitions(proj PRIVATE EXTERNAL_NLOHMANN_JSON)
-@@ -455,16 +454,14 @@ endif()
+@@ -455,16 +454,16 @@ endif()
```
Comment (SpaceIm) [inline recipes/proj/all/patches/0001-use-cmake-targets-9.0.1.patch:14]:
```suggestion
@@ -455,16 +454,14 @@ endif()
```

File: recipes/proj/all/patches/0001-use-cmake-targets-9.0.1.patch [hunk 15]
```diff
@@ -1,27 +1,30 @@
...
  
  if(TIFF_ENABLED)
    target_compile_definitions(proj PRIVATE -DTIFF_ENABLED)
 -  target_include_directories(proj PRIVATE ${TIFF_INCLUDE_DIR})
 -  target_link_libraries(proj PRIVATE ${TIFF_LIBRARY})
++  find_package(TIFF REQUIRED)
```
Comment (SpaceIm) [inline recipes/proj/all/patches/0001-use-cmake-targets-9.0.1.patch:20]:
```suggestion
```

File: recipes/proj/all/patches/0001-use-cmake-targets-9.0.1.patch [hunk 16]
```diff
@@ -1,27 +1,30 @@
...
 +  target_link_libraries(proj PRIVATE TIFF::TIFF)
  endif()
  
  if(CURL_ENABLED)
    target_compile_definitions(proj PRIVATE -DCURL_ENABLED)
 -  target_include_directories(proj PRIVATE ${CURL_INCLUDE_DIRS})
++  find_package(CURL REQUIRED)
```
Comment (SpaceIm) [inline recipes/proj/all/patches/0001-use-cmake-targets-9.0.1.patch:27]:
```suggestion
```

**Review by SpaceIm — CHANGES_REQUESTED** [review]
Simplification in 9.0.1 patch, needs to be done in all patches. Basically, except `CONAN_PKG::sqlite3` to replace by `SQLite::SQLite3`, there was no other modification needed in these patches (just maybe a find_package for nlohmann_json in old proj versions).

**Comment by paulharris** [comment]
@SpaceIm see changes to 6.x.x patch... you mean like this?
Should we specify MODULE or CONFIG in the find_package() calls?

File: recipes/proj/all/patches/0001-use-cmake-targets-7.2.1.patch [hunk 17]
```diff
@@ -1,24 +1,28 @@
...
 -  target_link_libraries(${PROJ_CORE_TARGET} ${TIFF_LIBRARY})
++  find_package(TIFF REQUIRED)
 +  target_link_libraries(${PROJ_CORE_TARGET} TIFF::TIFF)
  endif()
  
  if(CURL_ENABLED)
    target_compile_definitions(${PROJ_CORE_TARGET} PRIVATE -DCURL_ENABLED)
 -  target_include_directories(${PROJ_CORE_TARGET} PRIVATE ${CURL_INCLUDE_DIR})
 -  target_link_libraries(${PROJ_CORE_TARGET} ${CURL_LIBRARY})
++  find_package(CURL REQUIRED)
```
[the remark this answers is not in the document]
Comment (paulharris) [reply recipes/proj/all/patches/0001-use-cmake-targets-7.2.1.patch]: Thanks @SpaceIm, I've fixed up the scripts - please review :)

**Review by uilianries — APPROVED** [review]

**Pull request merged**